File: pcc/py_runtime/src/py_os_substrate.c
/* pcc/py_runtime/src/py_os_substrate.c
 *
 * C-only low-level OS primitives shared between the C and pcc-Python
 * runtime archives. Helpers here MUST live outside any module that
 * has a pcc-Python port in py/, because libpy_runtime_pcc_py.a swaps
 * those .o for the pcc-Python equivalents — anything pcc-Python
 * cannot express (struct stat layout, syscall buffer slabs, ...) has
 * to live here so both archives can link it.
 */
#include "py_internal.h"

#include <fcntl.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <time.h>
#include <unistd.h>

PyObject *py_time_monotonic(void) {
    return py_float_from_f64((double)pcc_runtime_monotonic_us() / 1.0e6);
}

PyObject *py_time_perf_counter(void) {
    return py_float_from_f64((double)pcc_runtime_monotonic_us() / 1.0e6);
}

PyObject *py_time_time(void) {
    return py_float_from_f64((double)pcc_runtime_now_us() / 1.0e6);
}

PyObject *py_time_strftime(PyObject *fmt) {
    PyObject *fmt_str = py_obj_str(fmt);
    if (fmt_str == NULL) return NULL;
    const char *raw_fmt = py_str_utf8(fmt_str);
    if (raw_fmt == NULL) {
        py_decref(fmt_str);
        return py_str_new("", 0);
    }
    time_t now = (time_t)(pcc_runtime_now_us() / 1000000);
    struct tm tmv;
#if defined(_WIN32)
    struct tm *tmp = localtime(&now);
    if (tmp == NULL) {
        py_decref(fmt_str);
        return py_str_new("", 0);
    }
    tmv = *tmp;
#else
    if (localtime_r(&now, &tmv) == NULL) {
        py_decref(fmt_str);
        return py_str_new("", 0);
    }
#endif
    char buf[256];
    size_t n = strftime(buf, sizeof(buf), raw_fmt, &tmv);
    py_decref(fmt_str);
    return py_str_new(buf, (int64_t)n);
}

PyObject *py_sys_stdin_readline(void) {
    char *line = NULL;
    size_t cap = 0;
    ssize_t n = getline(&line, &cap, stdin);
    if (n < 0) {
        free(line);
        return py_str_new("", 0);
    }
    PyObject *out = py_str_new(line, (int64_t)n);
    free(line);
    return out;
}

PyObject *py_os_urandom(PyObject *n_obj) {
    int overflow = 0;
    int64_t n = py_int_to_i64(n_obj, &overflow);
    if (overflow || n < 0) {
        py_raise_owned(py_exc_new(PY_EXC_VALUEERROR, "negative argument not allowed"));
        return NULL;
    }
    PyObject *out = py_bytes_new(NULL, n);
    if (out == NULL || n == 0) return out;
    char *data = ((PyBytesObject *)out)->data;
#if defined(__APPLE__) || defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__)
    arc4random_buf(data, (size_t)n);
    return out;
#else
    int fd = open("/dev/urandom", O_RDONLY);
    if (fd < 0) {
        py_decref(out);
        py_raise_owned(py_exc_new(PY_EXC_OSERROR, "could not open /dev/urandom"));
        return NULL;
    }
    int64_t pos = 0;
    while (pos < n) {
        ssize_t got = read(fd, data + pos, (size_t)(n - pos));
        if (got <= 0) {
            close(fd);
            py_decref(out);
            py_raise_owned(py_exc_new(PY_EXC_OSERROR, "could not read /dev/urandom"));
            return NULL;
        }
        pos += (int64_t)got;
    }
    close(fd);
    return out;
#endif
}

/* Classify a path by struct stat type:
 *   0 → stat() failed (missing or permission)
 *   1 → regular file
 *   2 → directory
 *   3 → other (symlink target, fifo, socket, ...)
 *
 * Used by py_os_path.{c,py} to implement isfile / isdir without
 * encoding the platform-specific struct stat layout in pcc-Python. */
int32_t py_path_stat_kind(const char *p) {
    struct stat st;
    if (p == (const char *)0) return 0;
    if (stat(p, &st) != 0) return 0;
    if (S_ISREG(st.st_mode)) return 1;
    if (S_ISDIR(st.st_mode)) return 2;
    return 3;
}

/* Return the path's last-modification time as IEEE-754 double seconds
 * since the Unix epoch — same shape as Python's os.path.getmtime
 * return value. NaN signals stat() failure (caller decides whether to
 * raise OSError or return NaN as-is). */
double py_path_stat_mtime(const char *p) {
    struct stat st;
    if (p == (const char *)0) return (double)0.0 / 0.0;
    if (stat(p, &st) != 0) return (double)0.0 / 0.0;
#if defined(__APPLE__)
    /* st_mtimespec is the canonical struct on Darwin. */
    return (double)st.st_mtimespec.tv_sec
         + (double)st.st_mtimespec.tv_nsec / 1.0e9;
#else
    return (double)st.st_mtim.tv_sec
         + (double)st.st_mtim.tv_nsec / 1.0e9;
#endif
}

int64_t py_path_stat_size(const char *p) {
    struct stat st;
    if (p == (const char *)0) return -1;
    if (stat(p, &st) != 0) return -1;
    return (int64_t)st.st_size;
}

/* Return the current working directory as a NUL-terminated cstring
 * pointer. Uses a thread-local static buffer so the caller does not
 * own the result — copy before another call clobbers it. NULL on
 * failure (e.g. cwd was deleted, ENOMEM). 8KB buffer accommodates
 * any reasonable POSIX path; longer paths fail rather than truncate. */
static __thread char _cwd_buf[8192];
const char *py_path_getcwd(void) {
    if (getcwd(_cwd_buf, sizeof(_cwd_buf)) == (const char *)0) {
        return (const char *)0;
    }
    return _cwd_buf;
}

/* Resolve PATH to a canonical absolute path (symlinks collapsed) via
 * realpath(3). Thread-local static buffer like py_path_getcwd: the caller
 * does not own the result and must copy before the next call. Returns NULL
 * when realpath(3) fails (e.g. a component does not exist, ENOENT), so the
 * caller can fall back to lexical normalization. 8KB exceeds PATH_MAX. */
static __thread char _realpath_buf[8192];
const char *py_path_realpath(const char *p) {
    if (p == (const char *)0) {
        return (const char *)0;
    }
    if (realpath(p, _realpath_buf) == (const char *)0) {
        return (const char *)0;
    }
    return _realpath_buf;
}

/* sys.platform value, picked at C compile time. Keeps the platform
 * string off CPython's runtime — matches Python's sys.platform format
 * (lower-case prefix without minor version). */
static const char *_sys_platform_cstr(void) {
#if defined(__APPLE__)
    return "darwin";
#elif defined(__linux__)
    return "linux";
#elif defined(_WIN32) || defined(_WIN64)
    return "win32";
#elif defined(__FreeBSD__)
    return "freebsd";
#else
    return "unknown";
#endif
}

static const char *_platform_machine_cstr(void) {
#if defined(__APPLE__) && defined(__aarch64__)
    return "arm64";
#elif defined(__aarch64__)
    return "aarch64";
#elif defined(_M_ARM64)
    return "ARM64";
#elif defined(__x86_64__) || defined(_M_X64)
    return "x86_64";
#elif defined(__i386__) || defined(_M_IX86)
    return "i386";
#elif defined(__arm__)
    return "arm";
#elif defined(__powerpc64__)
    return "ppc64";
#elif defined(__powerpc__)
    return "ppc";
#else
    return "";
#endif
}

PyObject *py_sys_platform_str(void) {
    const char *s = _sys_platform_cstr();
    int64_t n = 0;
    while (s[n] != '\0') n++;
    return py_str_new(s, n);
}

PyObject *py_platform_machine_str(void) {
    const char *s = _platform_machine_cstr();
    int64_t n = 0;
    while (s[n] != '\0') n++;
    return py_str_new(s, n);
}

PyObject *py_platform_release_str(void) {
    /* Only used by the fallback branch in
     * _host_target_triple_for_self_backend() when `cc -dumpmachine`
     * is unavailable. Keep this C-subset friendly; the normal self-host
     * path gets the precise Darwin/Linux triple from the compiler probe. */
    return py_str_new("0", 1);
}

/* Boxed `os.getcwd()` — same value as Python's os.getcwd. NULL on
 * error (matching CPython's OSError on stat failure; pcc's caller
 * gets a NULL it can null-check). */
PyObject *py_os_getcwd_str(void) {
    const char *p = py_path_getcwd();
    if (p == (const char *)0) return NULL;
    int64_t n = 0;
    while (p[n] != '\0') n++;
    return py_str_new(p, n);
}

/* `sys.path` — closed-world minimal, but preserve CPython's path[0]
 * contract for the owned entry modes.  Runtime package/import paths are
 * compiled into the closed-world artifact and are not published here. */
PyObject *py_sys_path_list(void) {
    PyObject *path0 = NULL;
    int32_t mode = py_program_mode();
    if (mode == 3 || mode == 4) {
        path0 = py_str_new("", 0);
    } else if (mode == 1) {
        const char *raw = py_program_argv(0);
        const char *resolved = raw != NULL ? py_path_realpath(raw) : NULL;
        if (resolved != NULL) {
            const char *slash = strrchr(resolved, '/');
            if (slash != NULL) {
                int64_t length = slash == resolved
                    ? 1
                    : (int64_t)(slash - resolved);
                path0 = py_str_new(resolved, length);
            }
        }
    }
    if (path0 == NULL) {
        path0 = py_os_getcwd_str();
    }
    if (path0 == NULL) {
        path0 = py_str_new("", 0);
    }
    PyObject *lst = py_list_new(0);
    py_list_append(lst, path0);
    py_decref(path0);
    return lst;
}

/* `os.access(path, mode)` — returns 1 if accessible, 0 otherwise.
 * `mode` is one of os.F_OK / R_OK / W_OK / X_OK (0 / 4 / 2 / 1).
 *
 * py_obj_str returns the same PyStr (with incref) when `path` is
 * already a string, so the helper does not need a type-check up
 * front; the unconditional incref/decref pair is cheap and keeps
 * the implementation portable across both archive variants. */
int32_t py_os_access(PyObject *path, int32_t mode) {
    if (path == NULL) return 0;
    PyObject *owned = py_obj_str(path);
    if (owned == NULL) return 0;
    const char *raw = py_str_utf8(owned);
    int ok = (raw != NULL && access(raw, (int)mode) == 0) ? 1 : 0;
    py_decref(owned);
    return ok;
}

int32_t py_os_write(int32_t fd, PyObject *data) {
    if (data == NULL) return -1;
    const char *ptr = NULL;
    int64_t len = 0;
    PyObject *curr = data;
    while (curr != NULL) {
        int32_t tag = py_type_of(curr);
        if (tag == PY_TYPE_STR) {
            ptr = py_str_utf8(curr);
            len = py_str_byte_len(curr);
            break;
        } else if (tag == PY_TYPE_BYTES) {
            ptr = ((PyBytesObject *)curr)->data;
            len = ((PyBytesObject *)curr)->byte_len;
            break;
        } else if (tag == PY_TYPE_BYTEARRAY) {
            ptr = ((PyByteArrayObject *)curr)->data;
            len = ((PyByteArrayObject *)curr)->byte_len;
            break;
        } else if (tag == PY_TYPE_MEMORYVIEW) {
            curr = pcc_gc_load_ptr(curr, &((PyMemoryViewObject *)curr)->base);
        } else {
            return -1;
        }
    }
    if (ptr == NULL || len < 0) return -1;
    ssize_t written = write((int)fd, ptr, (size_t)len);
    return (int32_t)written;
}

File: pcc/py_runtime/py/py_os_substrate.py
"""pcc-Python port of py_os_substrate.c.

The stat helpers delegate ABI-sensitive ``struct stat`` reads to
pcc.unsafe intrinsics. The runtime source stays portable and does not encode
Darwin/Linux field offsets directly.
"""

__pcc_runtime_port__ = True

from pcc.py_runtime.py.py_abi_constants import (
    PY_TYPE_BYTEARRAY,
    PY_TYPE_BYTES,
    PY_TYPE_INT,
    PY_TYPE_MEMORYVIEW,
    PY_TYPE_STR,
)

from pcc.extern import (
    extern,
    c_abi_export,
    c_double,
    c_int32,
    c_int64,
    c_ptr,
    c_size_t,
    c_void,
)
from pcc.unsafe import (
    cstr,
    define_global_ptr_null,
    free,
    global_load_ptr,
    global_store_ptr,
    is_tagged_int,
    load_i8,
    load_i32,
    load_i64,
    malloc,
    null,
    ptr_add,
    ptr_is_null,
    store_i64,
    strlen,
    target_platform_machine,
    target_sys_platform,
)

localtime_r = extern("localtime_r", (c_ptr, c_ptr), c_ptr)
strftime = extern("strftime", (c_ptr, c_size_t, c_ptr, c_ptr), c_size_t)
py_decref = extern("py_decref", (c_ptr,), c_void)
py_obj_str = extern("py_obj_str", (c_ptr,), c_ptr)
py_str_new = extern("py_str_new", (c_ptr, c_int64), c_ptr)
py_str_utf8 = extern("py_str_utf8", (c_ptr,), c_ptr)
py_list_new = extern("py_list_new", (c_int64,), c_ptr)
py_list_append = extern("py_list_append", (c_ptr, c_ptr), c_void)
py_float_from_f64 = extern("py_float_from_f64", (c_double,), c_ptr)
pcc_runtime_now_us = extern("pcc_platform_wall_time_us", (), c_int64)
pcc_runtime_monotonic_us = extern("pcc_platform_monotonic_us", (), c_int64)
py_str_byte_len = extern("py_str_byte_len", (c_ptr,), c_int64)
py_bytes_new = extern("py_bytes_new", (c_ptr, c_int64), c_ptr)
py_int_value_i64 = extern("py_int_value_i64", (c_ptr,), c_int64)
arc4random_buf = extern("arc4random_buf", (c_ptr, c_size_t), c_void)
read_sys = extern("pcc_platform_read", (c_int64, c_ptr, c_int64), c_int64)
write_sys = extern("pcc_platform_write", (c_int64, c_ptr, c_int64), c_int64)
pcc_platform_access = extern(
    "pcc_platform_access", (c_ptr, c_int64), c_int64
)
pcc_platform_getcwd = extern(
    "pcc_platform_getcwd", (c_ptr, c_int64), c_ptr
)
pcc_platform_stat_kind = extern(
    "pcc_platform_stat_kind", (c_ptr,), c_int64
)
pcc_platform_stat_mtime = extern(
    "pcc_platform_stat_mtime", (c_ptr,), c_double
)
pcc_platform_realpath = extern(
    "pcc_platform_realpath", (c_ptr, c_ptr, c_int64), c_ptr
)
pcc_gc_load_ptr = extern("pcc_gc_load_ptr", (c_ptr, c_ptr), c_ptr)
py_program_argv = extern("py_program_argv", (c_int64,), c_ptr)
py_program_mode = extern("py_program_mode", (), c_int32)


def _type_of(obj) -> int:
    if is_tagged_int(obj):
        return PY_TYPE_INT
    return load_i32(obj, 8)


define_global_ptr_null("py_path_getcwd_buf")
define_global_ptr_null("py_path_realpath_buf")


@c_abi_export("py_path_stat_kind")
def py_path_stat_kind(p) -> int:
    return pcc_platform_stat_kind(p)


@c_abi_export("py_path_stat_mtime")
def py_path_stat_mtime(p) -> float:
    return pcc_platform_stat_mtime(p)


@c_abi_export("py_time_monotonic")
def py_time_monotonic():
    return py_float_from_f64(pcc_runtime_monotonic_us() * 0.000001)


@c_abi_export("py_time_perf_counter")
def py_time_perf_counter():
    return py_float_from_f64(pcc_runtime_monotonic_us() * 0.000001)


@c_abi_export("py_time_time")
def py_time_time():
    return py_float_from_f64(pcc_runtime_now_us() * 0.000001)


@c_abi_export("py_time_strftime")
def py_time_strftime(fmt):
    fmt_str = py_obj_str(fmt)
    if ptr_is_null(fmt_str):
        return null()
    raw_fmt = py_str_utf8(fmt_str)
    if ptr_is_null(raw_fmt):
        py_decref(fmt_str)
        return py_str_new(null(), 0)
    now_buf = malloc(8)
    tm_buf = malloc(128)
    out_buf = malloc(256)
    if ptr_is_null(now_buf) or ptr_is_null(tm_buf) or ptr_is_null(out_buf):
        free(now_buf)
        free(tm_buf)
        free(out_buf)
        py_decref(fmt_str)
        return null()
    store_i64(now_buf, 0, pcc_runtime_now_us() // 1000000)
    if ptr_is_null(localtime_r(now_buf, tm_buf)):
        free(now_buf)
        free(tm_buf)
        free(out_buf)
        py_decref(fmt_str)
        return py_str_new(null(), 0)
    n: int = strftime(out_buf, 256, raw_fmt, tm_buf)
    out = py_str_new(out_buf, n)
    free(now_buf)
    free(tm_buf)
    free(out_buf)
    py_decref(fmt_str)
    return out


@c_abi_export("py_sys_stdin_readline")
def py_sys_stdin_readline():
    buf = malloc(4096)
    if ptr_is_null(buf):
        return null()
    pos: int = 0
    while pos < 4095:
        got: int = read_sys(0, ptr_add(buf, pos), 1)
        if got <= 0:
            break
        if load_i8(buf, pos) == 10:
            pos = pos + 1
            break
        pos = pos + 1
    out = py_str_new(buf, pos)
    free(buf)
    return out


@c_abi_export("py_os_urandom")
def py_os_urandom(n_obj):
    n: int = py_int_value_i64(n_obj)
    if n < 0:
        return null()
    out = py_bytes_new(null(), n)
    if ptr_is_null(out):
        return null()
    if n > 0:
        arc4random_buf(ptr_add(out, 24), n)
    return out


@c_abi_export("py_path_getcwd")
def py_path_getcwd():
    buf = global_load_ptr("py_path_getcwd_buf")
    if ptr_is_null(buf):
        buf = malloc(8192)
        if ptr_is_null(buf):
            return null()
        global_store_ptr("py_path_getcwd_buf", buf)
    result = pcc_platform_getcwd(buf, 8192)
    if ptr_is_null(result):
        return null()
    return buf


@c_abi_export("py_path_realpath")
def py_path_realpath(p):
    if ptr_is_null(p):
        return null()
    buf = global_load_ptr("py_path_realpath_buf")
    if ptr_is_null(buf):
        buf = malloc(8192)
        if ptr_is_null(buf):
            return null()
        global_store_ptr("py_path_realpath_buf", buf)
    result = pcc_platform_realpath(p, buf, 8192)
    if ptr_is_null(result):
        return null()
    return buf


def _str_from_cstr(p):
    if ptr_is_null(p):
        return null()
    return py_str_new(p, strlen(p))


@c_abi_export("py_sys_platform_str")
def py_sys_platform_str():
    return _str_from_cstr(target_sys_platform())


@c_abi_export("py_platform_machine_str")
def py_platform_machine_str():
    return _str_from_cstr(target_platform_machine())


@c_abi_export("py_platform_release_str")
def py_platform_release_str():
    return py_str_new(cstr("0"), 1)


@c_abi_export("py_os_getcwd_str")
def py_os_getcwd_str():
    return _str_from_cstr(py_path_getcwd())


@c_abi_export("py_sys_path_list")
def py_sys_path_list():
    path0 = null()
    mode: int = py_program_mode()
    if mode == 3 or mode == 4:
        path0 = py_str_new(null(), 0)
    elif mode == 1:
        raw = py_program_argv(0)
        if not ptr_is_null(raw):
            resolved = py_path_realpath(raw)
            if not ptr_is_null(resolved):
                i: int = 0
                last_slash: int = -1
                while load_i8(resolved, i) != 0:
                    if load_i8(resolved, i) == 47:
                        last_slash = i
                    i += 1
                if last_slash >= 0:
                    length: int = last_slash
                    if last_slash == 0:
                        length = 1
                    path0 = py_str_new(resolved, length)
    if ptr_is_null(path0):
        path0 = py_os_getcwd_str()
    if ptr_is_null(path0):
        path0 = py_str_new(null(), 0)
    lst = py_list_new(0)
    py_list_append(lst, path0)
    py_decref(path0)
    return lst


@c_abi_export("py_os_access")
def py_os_access(path, mode: int) -> int:
    if ptr_is_null(path):
        return 0
    owned = py_obj_str(path)
    if ptr_is_null(owned):
        return 0
    raw = py_str_utf8(owned)
    ok: int = 0
    if not ptr_is_null(raw):
        if pcc_platform_access(raw, mode) == 0:
            ok = 1
    py_decref(owned)
    return ok


@c_abi_export("py_os_write")
def py_os_write(fd: int, data) -> int:
    if ptr_is_null(data):
        return -1
    ptr = null()
    length = 0
    curr = data
    while not ptr_is_null(curr):
        tag = _type_of(curr)
        if tag == PY_TYPE_STR:  # PY_TYPE_STR
            ptr = py_str_utf8(curr)
            length = py_str_byte_len(curr)
            break
        elif tag == PY_TYPE_BYTES:  # PY_TYPE_BYTES
            ptr = ptr_add(curr, 24)
            length = load_i64(curr, 16)
            break
        elif tag == PY_TYPE_BYTEARRAY:  # PY_TYPE_BYTEARRAY
            ptr = ptr_add(curr, 24)
            length = load_i64(curr, 16)
            break
        elif tag == PY_TYPE_MEMORYVIEW:  # PY_TYPE_MEMORYVIEW
            # base pointer is at offset 16
            curr = pcc_gc_load_ptr(curr, ptr_add(curr, 16))
        else:
            return -1

    if ptr_is_null(ptr) or length < 0:
        return -1
    return write_sys(fd, ptr, length)
